Modify rules S6185,S6484,S6494,S6495: RSPEC improvements (CPP-5056) (#3816)
SonarSource/rspec: rules/S6494/cfamily/rule.adoc

In contrast to C printf-like functions, {cpp} provides safer and more robust interfaces for performing text formatting:

* The `std::format` interface family ({cpp}20) allows formatting text into a string.
* The `std::print` interface family ({cpp}23) allows printing formatted text.

> Firstly, {cpp} formatting facilities perform validation of the format string against the type of the formatted argument. If the validation fails, it is reported as a compilation error for the calls of `std::print` and `std::format`.
> In addition, `std::vformat`, `std::vprint_unicode`, and `std::vprint_nonunicode` report validation failures by throwing an instance of `std::format_error`.

Reviewer: Not sure of the "Firstly" "In addition" here. 
```suggestion
{cpp} formatting facilities perform validation of the format string against the type of the formatted argument. If the validation fails, it is reported as a compilation error for the calls of `std::print` and `std::format`. `std::vformat`, `std::vprint_unicode`, and `std::vprint_nonunicode`, that can't validate at compile-time, report failures by throwing an instance of `std::format_error`.
```

Author: Sounds better, indeed. One nit though: "that can't validate at at compile-time" --> it's a design choice. I would rephrase it a bit:

```suggestion
{cpp} formatting facilities perform validation of the format string against the type of the formatted argument. If the validation fails, it is reported as a compilation error for the calls of `std::print` and `std::format`. When the format string is not available at compile-time, `std::vformat`, `std::vprint_unicode`, and `std::vprint_nonunicode` can be used. They will report failures at runtime by throwing an instance of `std::format_error`.
```

WDYT?

Reviewer: LGTM

[source,cpp]
----

> void printTextIntoBuffer(char* out) {

Reviewer: ```suggestion
void printTextIntoBuffer(char* out) { // Assumes the buffer pointed-to by out is large enough
```

Author: Makes sense. I've added the comment on a separate line to keep them short. ad02ba96e8

  sprintf(out, "%u %s", 10u, "text"); // Noncompliant
}

void printTextIntoSizedBuffer(char* out, size_t n) {
  std::snprintf(out, n, "%i %% %LG", 10, 10.0L); // Noncompliant
}

void printToFile(FILE* f) {
  printf("%i", 10); // Noncompliant since C++23
  std::fprintf(f, "%f", 10.0); // Noncompliant since C++23
}
----

=== Compliant solution

[source,cpp]
----

> void printTextIntoBuffer(char* out) {

Reviewer: ```suggestion
void printTextIntoBuffer(char* out) { // Assumes the buffer pointed-to by out is large enough
```

* {cpp} reference - https://en.cppreference.com/w/cpp/io/vprint_nonunicode[`std::vprint_nonunicode`]
* {cpp} reference - https://en.cppreference.com/w/cpp/io/basic_ostream/vprint_unicode[`std::vprint_unicode(std::ostream)`]

> * {cpp} reference - https://en.cppreference.com/w/cpp/io/basic_ostream/vprint_nonunicode[`std::vprint_nonunicode(std::ostream)`]

Reviewer: Why only those 3? They don't seem the most relevant...
Maybe a global link to https://en.cppreference.com/w/cpp/utility/format

Author: The last two are from `<ostream>` and are not available from the page you linked.
I'm not sure how important that is, though. Maybe we should have fewer links...

What about a link to:
 * std::format
 * std::print
 * std::print(ostream)
 * <format>
 * <print>

and we "ignore" the rest here, assuming the reader can navigate to them?

Reviewer: I think we need `format_to`, because we use it so much in the example... (or should we change the example? For instance by returning a string?
I'm not sure of the value of having both  `std::format` and `<format>`, it's easy to navigate from one to thee other.

Author: - links: std::print(ostream) `<format>` `<print>`
- update example: keep same interface with format_to_n, + alternative with format

Author: 1. done.
2. re-reading the examples, I found more issues. I split them into smaller functions and gave one example of redesign.
